rewrite-core/src/test/java/org/openrewrite/remote/RemoteArchiveTest.java

```diff
@@ -48,15 +56,50 @@ void gradleWrapper(String version) throws Exception {
           )
           .build("gradle-[^\\/]+\\/(?:.*\\/)+gradle-wrapper-(?!shared).*\\.jar");
 
-        byte[] actual = readAll(remoteArchive.getInputStream(ctx));
-        assertThat(actual).hasSizeGreaterThan(50_000);
+        long actual = getInputStreamSize(remoteArchive.getInputStream(ctx));
+        assertThat(actual).isGreaterThan(50_000);
     }
 
-    private byte[] readAll(InputStream is) {
+    @ParameterizedTest
+    @ValueSource(strings = {"7.4.2", "7.5-rc-1", "7.6"})
+    void gradleWrapperConcurrent(String version) throws Exception {
+        int executionCount = 5;
+        ExecutorService executorService = Executors.newFixedThreadPool(executionCount);
+        CompletionService<Long> completionService = new ExecutorCompletionService<>(executorService);
+
+        for (int i = 0; i < executionCount; i++) {
+            completionService.submit(() -> {
+                URL distributionUrl = requireNonNull(RemoteArchiveTest.class.getClassLoader()
+                  .getResource("gradle-" + version + "-bin.zip"));
+
+                ExecutionContext ctx = new InMemoryExecutionContext();
+                HttpSenderExecutionContextView.view(ctx)
+                  .setLargeFileHttpSender(new MockHttpSender(distributionUrl::openStream));
+
+                RemoteArchive remoteArchive = Remote
+                  .builder(
+                    Paths.get("gradle/wrapper/gradle-wrapper.jar"),
+                    distributionUrl.toURI()
+                  )
+                  .build("gradle-[^\\/]+\\/(?:.*\\/)+gradle-wrapper-(?!shared).*\\.jar");
+
+                return getInputStreamSize(remoteArchive.getInputStream(ctx));
+            });
+        }
+
+        for (int i = 0; i < executionCount; i++) {
+            Future<Long> result = completionService.take();
+            Long actual = result.get();
+            assertThat(actual).isGreaterThan(50_000);
+        }
+
+        executorService.shutdown();
+    }
+
+    private Long getInputStreamSize(InputStream is) {
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
```

Comment:
@natedanner This doesn't help the heap problem, a ByteArrayOutputStream is still backed by a byte array.

Reply:
Do what @shanman190  suggested and use the `CountingOutputStream`

Reply:
Sorry, this was meant to be the noop output stream suggested by @shanman190. Replaced.

Reply:
I think this should be resolved now that we're tossing the bytes away.